overlord/devicestate/handlers.go

```diff
@@ -58,56 +60,83 @@ func useStaging() bool {
 	}
-	return "https://api.snapcraft.io/"
+	return mustParse("https://api.snapcraft.io/")
+}
+
+func mustParse(s string) *url.URL {
+	u, err := url.Parse(s)
+	if err != nil {
+		panic(err)
+	}
+	return u
 }
 
 var (
 	keyLength     = 4096
 	retryInterval = 60 * time.Second
 	maxTentatives = 15
-	baseStoreURL  = baseURL()
+	baseStoreURL  = baseURL().ResolveReference(authRef)
+
+	authRef    = mustParse("api/v1/snaps/auth/") // authRef must end in / for the following refs to work
+	reqIdRef   = mustParse("request-id")
+	serialRef  = mustParse("serial")
+	devicesRef = mustParse("devices")
 )
 
-func deviceAPIbase(proxyStore *asserts.Store, svcURL string) (s string) {
-	if proxyStore != nil && svcURL != "" {
-		// TODO: warning
-		logger.Noticef("UNSUPPORTED! trying to use a snap store proxy (at %q) together with a device service URL (%q). Falling back to contacting the service URL directly.", proxyStore.URL(), svcURL)
-	}
-	if svcURL != "" {
-		if svcURL[len(svcURL)-1] != '/' {
-			svcURL += "/"
-		}
-		return svcURL
-	}
+func newEnoughProxy(proxyURL *url.URL, client *http.Client) bool {
+	const prefix = "Cannot check whether proxy store supports a custom serial vault"
 
-	var base string
-	if proxyStore != nil {
-		base = proxyStore.URL().String()
+	req, err := http.NewRequest("HEAD", proxyURL.String(), nil)
+	if err != nil {
+		// can't really happen unless proxyURL is somehow broken
+		logger.Debugf(prefix+": %v", err)
+		return false
+	}
+	req.Header.Set("User-Agent", httputil.UserAgent())
+	resp, err := client.Do(req)
+	if err != nil {
+		// some sort of network or protocol error
+		logger.Debugf(prefix+": %v", err)
+		return false
 	}
-	if base == "" {
-		base = baseStoreURL
+	resp.Body.Close()
+	if resp.StatusCode != 200 {
+		logger.Debugf(prefix+": Head request returned %s.", resp.Status)
+		return false
 	}
-	if base[len(base)-1] != '/' {
-		base += "/"
+	verstr := resp.Header.Get("Snap-Store-Version")
+	ver, err := strconv.Atoi(verstr)
+	if err != nil {
+		logger.Debugf(prefix+": Bogus Snap-Store-Version header %q.", verstr)
+		return false
 	}
-
-	return base + "api/v1/snaps/auth/"
+	return ver >= 6
 }
 
-func requestIDURL(proxyStore *asserts.Store, svcURL string) string {
-	return deviceAPIbase(proxyStore, svcURL) + "request-id"
-}
+func (cfg *serialRequestConfig) setURLs(proxyURL, svcURL *url.URL) {
+	base := baseStoreURL
+	if proxyURL != nil {
+		if svcURL != nil {
+			if cfg.headers == nil {
+				cfg.headers = make(map[string]string, 1)
+			}
+			cfg.headers["X-Snap-Device-Service-URL"] = svcURL.String()
+		}
+		base = proxyURL.ResolveReference(authRef)
+	} else if svcURL != nil {
+		base = svcURL
+	}
 
-func serialRequestURL(proxyStore *asserts.Store, svcURL string) string {
-	base := deviceAPIbase(proxyStore, svcURL)
-	if svcURL == "" {
-		return base + "devices"
+	cfg.requestIDURL = base.ResolveReference(reqIdRef).String()
+	if svcURL != nil && proxyURL == nil {
+		cfg.serialRequestURL = base.ResolveReference(serialRef).String()
```

Comment:
can we add some comment here, like "talking directly to a custom device service" ?